src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java

```diff
@@ -530,31 +536,66 @@ private Result shouldSaveResult(@NotNull Result result, boolean shouldSave) {
         Set<ProgrammingExerciseTask> tasks = programmingExerciseTaskService.getTasksWithUnassignedTestCases(exerciseId);
+
         List<FeedbackDetailDTO> feedbackDetails = studentParticipationRepository.findAggregatedFeedbackByExerciseId(exerciseId);
+        String searchTerm = search.getSearchTerm() != null ? search.getSearchTerm().toLowerCase() : "";
 
-        return feedbackDetails.stream().map(detail -> {
+        feedbackDetails = feedbackDetails.stream().map(detail -> {
             double relativeCount = (detail.count() * 100.0) / distinctResultCount;
-            int taskNumber = tasks.stream().filter(task -> task.getTestCases().stream().anyMatch(tc -> tc.getTestName().equals(detail.testCaseName()))).findFirst()
-                    .map(task -> tasks.stream().toList().indexOf(task) + 1).orElse(0);
+            int taskNumber = determineTaskNumberOfTestCase(detail.testCaseName(), tasks);
+
             return new FeedbackDetailDTO(detail.count(), relativeCount, detail.detailText(), detail.testCaseName(), taskNumber);
-        }).toList();
+        }).filter(matchesSearchTerm(searchTerm)).sorted(getComparatorForFeedbackDetails(search)).toList();
+
+        return paginateFeedbackDetails(feedbackDetails, search.getPage(), search.getPageSize());
+    }
+
+    private Predicate<FeedbackDetailDTO> matchesSearchTerm(String searchTerm) {
+        return detail -> searchTerm.isEmpty() || detail.detailText().toLowerCase().contains(searchTerm) || detail.testCaseName().toLowerCase().contains(searchTerm)
+                || String.valueOf(detail.count()).contains(searchTerm) || String.valueOf(detail.taskNumber()).contains(searchTerm)
+                || String.valueOf(detail.relativeCount()).contains(searchTerm);
+    }
+
+    private Comparator<FeedbackDetailDTO> getComparatorForFeedbackDetails(SearchTermPageableSearchDTO<String> search) {
+        Map<String, Comparator<FeedbackDetailDTO>> comparators = Map.of("count", Comparator.comparingLong(FeedbackDetailDTO::count), "detailText",
+                Comparator.comparing(FeedbackDetailDTO::detailText, String.CASE_INSENSITIVE_ORDER), "testCaseName",
+                Comparator.comparing(FeedbackDetailDTO::testCaseName, String.CASE_INSENSITIVE_ORDER), "taskNumber", Comparator.comparingInt(FeedbackDetailDTO::taskNumber),
+                "relativeCount", Comparator.comparingDouble(FeedbackDetailDTO::relativeCount));
+        Comparator<FeedbackDetailDTO> comparator = comparators.getOrDefault(search.getSortedColumn(), (a, b) -> 0);
+        return search.getSortingOrder() == SortingOrder.ASCENDING ? comparator : comparator.reversed();
+    }
+
+    private int determineTaskNumberOfTestCase(String testCaseName, Set<ProgrammingExerciseTask> tasks) {
+        return IntStream.range(0, tasks.size()).filter(i -> tasks.stream().toList().get(i).getTestCases().stream().anyMatch(tc -> tc.getTestName().equals(testCaseName)))
```

Comment:
I don't think that this will work in all cases. Please also check what happens if you later add a new task in the programming exercise either at the end or somewhere between other existing tasks.